cpp/src/lists/segmented_sort.cu

```diff
@@ -163,7 +163,11 @@ struct SegmentedSortColumn {
     // the average list size at which to prefer radixsort:
     constexpr cudf::size_type MIN_AVG_LIST_SIZE_FOR_RADIXSORT{100};
 
-    if ((child.size() / offsets.size()) < MIN_AVG_LIST_SIZE_FOR_RADIXSORT) {
+    // Floating point types are disqualified from radix sort because the
+    // bit-sorting algorithm in CUB places -NaN away from NaN which is
+    // not expected in cudf (issue 11703)
+    if (std::is_floating_point_v<T> ||
```

Comment:
We should handle this at a higher level with SFINAE -- we already use `segmented_sort_by_key` for non-numeric types, and it's awkward to combine type checking and size heuristics inline here when we already use SFINAE for type constraints. We should just modify the SFINAE to use a trait like `numeric and not floating`, or perhaps even `representation type is integral`.